akka-stream/src/main/scala/akka/stream/impl/fusing/Interpreter.scala

```diff
@@ -282,18 +319,48 @@ private[akka] class OneBoundedInterpreter(ops: Seq[Stage[_, _]], val forkLimit:
       null
     }
 
-    override def hold(): FreeDirective = {
-      if (currentOp.holding) throw new IllegalStateException("Cannot hold while already holding")
-      currentOp.holding = true
+    override def holdUpstream(): UpstreamDirective = {
+      removeBits(PrecedingWasPull)
+      addBits(UpstreamBall)
       exit()
     }
 
-    override def isHolding: Boolean = currentOp.holding
+    override def holdUpstreamAndPush(elem: Any): UpstreamDirective = {
+      ReactiveStreamsCompliance.requireNonNullElement(elem)
```

Comment:
The real conflict then arises not from the engine itself but from the usage of `Stage` and `Context` for the user-facing API (in the context of a Reactive Streams compliant implementation) as well as the internal implementation. At this point I see no reason to complicate matters for our own use, it should be rather simple to disentangle (but quite a bit of extra code) once we need these to be independent.